docs/callbacks.rst

```diff
@@ -17,80 +22,359 @@ on abstraction of Browser/Server communication. Here is a simple syntax in Agile
         return $action->text(rand(1,100));
     });
 
-For comparison, the code without PHP callback would look like this::
+This creates call-back route transparently which is triggered automatically during the 'click' event.
+To make this work seamlessly there are several classes at play. This documentation chapter will walk
+you through the callback mechanisms of Agile UI.
 
-    $button = new Button();
+The Callback class
+------------------
 
-    // clicking button will change text only once, because
-    // random is generated during render and stored statically.
-    $button->on('click')->text(rand(1,100));
+.. php:class:: Callback
 
+Callback is not a View. This class does not extend any other class but it does implement several important
+traits:
 
-Lets spend a moment to discuss how this actually works and why. Agile UI works under the following
-assumptions:
+ - TrackableTrait [todo add link]
+ - AppScopeTrait
+ - DIContainerTrait
 
- - Sending same GET request will execute the Button creation code (as displayed above) once again.
- - Adding additional GET parameters will not affect results of GET request, assuming that those arguments don't conflict.
- - URL is in format <...>?param=1&param2=2
- - Executing "echo" followed by "exit" will not be punished by the environment.
+To create a new callback, do this::
 
-If any of the above are false inside your environment (e.g. you use custom output buffering), you will have to
-fine-tune Agile Data Application class to work with your environment requirement.
+    $c = new \atk4\ui\Callback();
+    $layout->add($c);
 
-From the above assumptions, Agile UI derrives at conclusion, that by executing AJAX request from the
-browser with a special GET argument a certain code can be triggered inside the on() method that short-circuts
-output producing response specifically for button.
+Adding Callback into any object will not affect the rendering, but it will make Callback part of the
+:ref:`render_tree` and it will produce it with a unique callback URL:
 
-Callback Logic Explained
-------------------------
+.. php:method:: getURL($val)
 
-1. When your code include a call-back format, an instance of '\atk4\ui\Callback' is created and added into
-the render-tree.
+.. php:method:: set
 
-2. As any other object in Render-tree, Callback receives a unique name.
+The following example code generates unique URL::
 
-3. Callback asks Application class to build a new URL based on the current arguments plus a new GET argument that it
-will be able to recognize, typically &my_full_name=callback
+    $label = $layout->add(['Label','Callback URL:']);
+    $cb = $label->add('Callback');
+    $label->detail = $cb->getURL();
+    $label->link($cb->getURL());
```

Comment:
Maybe one more sentence about what we will see on screen and what can we do with it.